relayer/app/cursor/stream.go

```diff
@@ -0,0 +1,160 @@
+		if c.GetConfirmed() {
+			return c.GetAttestOffset(), true, nil
+		}
+	}
+
+	return 0, false, nil
+}
+
+// getAllCursors returns all cursors from the storage.
+//
+// Cursors are ordered by source chain ID, confirmation level and destination chain ID in ascending order.
+func (s *StreamCursors) getAllCursors(ctx context.Context) ([]*Cursor, error) {
+	iterator, err := s.cursors.List(ctx,
+		CursorPrimaryKey{}.WithSrcChainIdConfLevelDstChainId(
+			s.sourceChainID,
+			uint32(s.confLevel),
+			s.destChainID,
+		))
+
+	if err != nil {
+		return nil, errors.Wrap(err, "list cursors")
+	}
+	defer iterator.Close()
+
+	var cursors []*Cursor
+	for iterator.Next() {
+		cursor, err := iterator.Value()
+		if err != nil {
+			return nil, errors.Wrap(err, "cursor value")
+		}
+		cursors = append(cursors, cursor)
+	}
+
+	return cursors, nil
+}
+
+// Confirm checks all stream cursors and confirms them according to the rules:
+//   - if a cursor is not empty fetch finalized cursor from the network and confirm only if finalized
+//     message offset is higher to the cursor last message offset
+//   - if a cursor is empty, confirm it if the previous cursor was confirmed
+func (s *StreamCursors) Confirm(ctx context.Context) error {
+	cursors, err := s.getAllCursors(ctx)
+	if err != nil {
+		return err
+	}
+
+	// if we only have empty and not confirmed skip as there's nothing to do
+	nonEmptyOrConfirmed := slices.ContainsFunc(cursors, func(c *Cursor) bool {
+		return !c.GetEmpty() || c.GetConfirmed()
+	})
+	if !nonEmptyOrConfirmed {
+		return nil
+	}
+
+	for _, cursor := range cursors {
+		// if cursor not empty, confirm it comparing to the network finalized cursor
+		if !cursor.GetEmpty() && !cursor.GetConfirmed() {
+			confirmed, err := s.isFinalized(ctx, cursor)
+			if err != nil {
+				return err
+			} else if !confirmed {
+				// we reached latest confirmed
+				break
+			}
+		}
+
+		cursor.Confirmed = true
+		if err := s.cursors.Update(ctx, cursor); err != nil {
+			return errors.Wrap(err, "cursor update")
+		}
+	}
+
+	return nil
+}
+
+// Trim iterates over all stored stream cursors in reverse order and when it finds the first
+// confirmed it leaves it but deletes all previous cursors as they are confirmed.
+func (s *StreamCursors) Trim(ctx context.Context) error {
+	cursors, err := s.getAllCursors(ctx)
+	if err != nil {
+		return err
+	}
+
+	var confirmed bool
+	slices.Reverse(cursors)
+	for _, c := range cursors {
+		// we iterate until we find first confirmed, skip first then delete all after
+		if !confirmed {
+			confirmed = c.GetConfirmed()
+			continue
+		}
+
+		if err := s.cursors.Delete(ctx, c); err != nil {
+			return errors.Wrap(err, "delete cursor")
+		}
+	}
+
+	return nil
+}
+
+// isFinalized checks if the cursor is finalized on the network.
+func (s *StreamCursors) isFinalized(ctx context.Context, cursor *Cursor) (bool, error) {
+	final, ok, err := s.provider.GetSubmittedCursor(ctx, xchain.FinalizedRef, cursor.StreamID())
```

Comment:
Conf level to shard ID conversion, which I'm not sure about, but since we are not persisting shards now I wonder if that is ok